File: core/src/main/java/pcb/uwu/amount/derived/thermodynamics/Pressure.java
package pcb.uwu.amount.derived.thermodynamics;

import org.jetbrains.annotations.NotNull;
import pcb.uwu.core.BigDecimalAmount;
import pcb.uwu.core.CompositeUnitAmount;
import pcb.uwu.core.Magnitude;
import pcb.uwu.core.UnitAmount;
import pcb.uwu.unit.derived.termodynamics.PressureUnit;

import java.math.BigDecimal;
import java.math.MathContext;

import static pcb.uwu.utils.UnitAmountUtils.dividedByScalar;
import static pcb.uwu.utils.UnitAmountUtils.getAmountIn;
import static pcb.uwu.utils.UnitAmountUtils.minusAmount;
import static pcb.uwu.utils.UnitAmountUtils.multipliedByScalar;
import static pcb.uwu.utils.UnitAmountUtils.plusAmount;

public class Pressure extends CompositeUnitAmount<PressureUnit> {

	// region constructors

	public Pressure(Number value, PressureUnit unit) {
		super(value, unit);
	}

	public Pressure(Number value, Magnitude magnitude, PressureUnit unit) {
		super(value, magnitude, unit);
	}

	public Pressure(String value, PressureUnit unit) {
		super(value, unit);
	}

	public Pressure(String value, Magnitude magnitude, PressureUnit unit) {
		super(value, magnitude, unit);
	}

	public Pressure(BigDecimal value, PressureUnit unit) {
		super(value, unit);
	}

	public Pressure(BigDecimal value, Magnitude magnitude, PressureUnit unit) {
		super(value, magnitude, unit);
	}

	public Pressure(BigDecimalAmount amount, PressureUnit unit) {
		super(amount, unit);
	}

	public Pressure(BigDecimalAmount amount, Magnitude magnitude, PressureUnit unit) {
		super(amount, magnitude, unit);
	}

	// endregion

	// region implement UnitAmount

	@NotNull
	@Override
	public Pressure plus(@NotNull UnitAmount<PressureUnit> other) {
		return new Pressure(plusAmount(this, other), getUnit());
	}

	@NotNull
	@Override
	public Pressure minus(@NotNull UnitAmount<PressureUnit> other) {
		return new Pressure(minusAmount(this, other), getUnit());
	}

	@Override
	public Pressure times(BigDecimal other, MathContext mathContext) {
		return new Pressure(multipliedByScalar(this, other, mathContext), getUnit());
	}

	@Override
	public Pressure div(BigDecimal other, MathContext mathContext) {
		return new Pressure(dividedByScalar(this, other, mathContext), getUnit());
	}

	@Override
	public Pressure into(PressureUnit unit) {
		return new Pressure(getAmountIn(this, unit), unit);
	}

	// endregion
}

File: core/src/main/java/pcb/uwu/amount/derived/thermodynamics/Pressure.kt
package pcb.uwu.amount.derived.thermodynamics

import pcb.uwu.core.CompositeUnitAmount
import pcb.uwu.core.Magnitude
import pcb.uwu.core.Magnitude.NATURAL
import pcb.uwu.core.UnitAmount
import pcb.uwu.unit.derived.termodynamics.PressureUnit
import pcb.uwu.utils.UnitAmountUtils

open class Pressure : CompositeUnitAmount<PressureUnit>
{
    @JvmOverloads
    constructor(amount: Number,
                magnitude: Magnitude = NATURAL,
                unit: PressureUnit)
            : super(amount, magnitude, unit)

    @JvmOverloads
    constructor(amount: String,
                magnitude: Magnitude = NATURAL,
                unit: PressureUnit)
            : super(amount, magnitude, unit)

    // region UnitAmount

    override fun plus(amount: UnitAmount<PressureUnit>) =
        Pressure(amount = this.amount + (amount to this.unit).amount,
                 unit = this.unit)

    override fun minus(amount: UnitAmount<PressureUnit>) =
        Pressure(amount = this.amount - (amount to this.unit).amount,
                 unit = this.unit)

    override fun times(number: Number) =
        Pressure(amount = this.amount * number,
                 unit = this.unit)

    override fun div(number: Number) =
        Pressure(amount = this.amount / number,
                 unit = this.unit)

    override fun to(unit: PressureUnit) =
        Pressure(amount = UnitAmountUtils.getAmountIn(unitAmount = this, newUnit = unit),
                 unit = unit)

    // endregion
}